airflow/models/dag.py

```diff
@@ -1941,6 +2087,36 @@ def deactivate_deleted_dags(cls, alive_dag_filelocs: List[str], session=None):
             session.rollback()
             raise
 
+    @classmethod
+    def dags_needing_dagruns(cls, session: Session):
+        """
+        Return (and lock) a list of Dag objects that are due to create a new DagRun This will return a
+        resultset of rows  that is row-level-locked with a "SELECT ... FOR UPDATE" query, you should ensure
+        that any scheduling decisions are made in a single transaction -- as soon as the transaction is
+        committed it will be unlocked.
+        """
+
+        # TODO[HA]: Bake this query, it is run _A lot_
+        # TODO[HA]: Make this limit a tunable. We limit so that _one_ scheduler
+        # doesn't try to do all the creation of dag runs
+        return session.query(cls).filter(
+            cls.is_paused.is_(False),
+            cls.is_active.is_(True),
+            cls.next_dagrun_create_after <= func.now(),
+        ).order_by(
+            cls.next_dagrun_create_after
+        ).limit(10).with_for_update(of=cls, **skip_locked(session=session))
```

Comment:
```python
		        # TODO[HA]: Make this limit a tunable. We limit so that _one_ scheduler
		        # doesn't try to do all the creation of dag runs
```

Reply:
The main reason for limiting this is two fold:

1. so that when running multiple schedulers, one scheduler doesn't try to do _all_ the work of creating dag runs.
2. when running with a single scheduler, make sure that the scheduler doesn't get "stuck" just creating dag runs instead of scheduling some tasks.

Reply:
Changed to tunable in https://github.com/apache/airflow/pull/10956/commits/ef9101ff910eba85a2782f894f0e9dfbafc04689